Source/WebCore/page/csp/ContentSecurityPolicy.cpp

```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
     violationEventInit.effectiveDirective = effectiveViolatedDirective;
     violationEventInit.originalPolicy = violatedDirectiveList.header();
     violationEventInit.sourceFile = info.sourceFile;
+    // WPT expects this behavior:
+    if (usesReportTo && violationEventInit.sourceFile.isNull())
+        violationEventInit.sourceFile = info.documentURI;
     violationEventInit.disposition = violatedDirectiveList.isReportOnly() ? SecurityPolicyViolationEventDisposition::Report : SecurityPolicyViolationEventDisposition::Enforce;
     violationEventInit.statusCode = httpStatusCode;
     violationEventInit.lineNumber =  info.lineNumber;
     violationEventInit.columnNumber = info.columnNumber;
     violationEventInit.sample = info.sample;
     violationEventInit.bubbles = true;
     violationEventInit.composed = true;
+
+    Vector<String> reportURIs;
+    if (usesReportTo) {
+        if (m_reportingClient) {
+            m_reportingClient->notifyReportObservers(Report::create(CSPViolationReportBody::cspReportType(), info.documentURI, CSPViolationReportBody::create(SecurityPolicyViolationEventInit { violationEventInit })));
+
+            for (auto& token : violatedDirectiveList.reportToTokens()) {
+                auto reportToURI = m_reportingClient->endpointForToken(token);
+                if (reportToURI.isNull() || reportToURI.isEmpty())
+                    continue;
+                
+                reportURIs.append(WTFMove(reportToURI));
+            }
+        } else {
+            auto& reporting = reportingScopeForContext(*m_scriptExecutionContext);
+            reporting.notifyReportObservers(Report::create(CSPViolationReportBody::cspReportType(), info.documentURI, CSPViolationReportBody::create(SecurityPolicyViolationEventInit { violationEventInit })));
+            
+            for (auto& token : violatedDirectiveList.reportToTokens()) {
```

Comment:
It's a bit sad that we're duplicating almost identical code here.
Maybe we can use template function to share code?
Alternatively, maybe ReportingScope should implement ReportingClient interface.

Reply:
I think I've cleaned this up now.